### ISSUE TITLE
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'

### ISSUE THREAD

**Issue by c01danjo** [PowerShell/PowerShell#7909]
Hi,
I am using Powershell SDK to execute powershell commands from an ASP.net core 2.1 application. It works in in dev (when i run it from visual studio). But when i deploy to IIS in production i keep getting: 

System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.
File name: 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Management.Automation.Runspaces.PSSnapInHelpers.GetAssemblyTypes(Assembly assembly, String name)
   at System.Management.Automation.Runspaces.PSSnapInHelpers.AnalyzeModuleAssemblyWithReflection(Assembly assembly, String name, PSSnapInInfo psSnapInInfo, PSModuleInfo moduleInfo, String helpFile, Dictionary`2& cmdlets, Dictionary`2& aliases, Dictionary`2& providers)
   at System.Management.Automation.Runspaces.PSSnapInHelpers.AnalyzePSSnapInAssembly(Assembly assembly, String name, PSSnapInInfo psSnapInInfo, PSModuleInfo moduleInfo, Dictionary`2& cmdlets, Dictionary`2& aliases, Dictionary`2& providers, String& helpFile)
   at System.Management.Automation.Runspaces.InitialSessionState.ImportPSSnapIn(PSSnapInInfo psSnapInInfo, PSSnapInException& warning)
   at System.Management.Automation.Runspaces.InitialSessionState.CreateDefault()
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace(PSHost host)
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace()
   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)

It works if I call Invoke() from a Console application. Any Idea why?

**Comment by bergmeister** [comment]
That sounds like a bug to me, the SDK Nuget package seems to have a missing required dependency to the [Microsoft.Management.Infrastructure](https://www.nuget.org/packages/Microsoft.Management.Infrastructure/) NuGet package as it can be seen below in the dependencies of the SDK Nuget package. You can manually add the `Microsoft.Management.Infrastructure` NuGet package, which should resolve your problem.

![image](https://user-images.githubusercontent.com/9250262/46291402-43e26480-c586-11e8-9df6-28b09bb5926e.png)

**Comment by c01danjo** [comment]
It seems that the problem i is Hangfire. When I start the code from the main tread it works as intended.

**Comment by Eyadius** [comment]
I have the same Problem, if start the Program in the Visual Studio it works fine, but if i host the Program as Windows Service it fires the same Exception. PLease help me

**Comment by iSazonov** [comment]
Perhaps #7886 helps.

**Comment by c01danjo** [comment]
Thank you  #7886 worked. I changed my Publish Profile to <RuntimeIdentifier>win10-x64</RuntimeIdentifier> and it worked fine.

**Comment by c01danjo** [comment]
Thank you #7886 worked!

**Comment by RaphiStein** [comment]
@c01danjo can you explain what you did exactly, if you can still recall?

**Comment by c01danjo** [comment]
> @c01danjo can you explain what you did exactly, if you can still recall?

I changed the property "RuntimeIdentifier" to "win10-x64" in the publishing configuration. /Properties/{your publishing configuration}.pubxml.

**Comment by RaphiStein** [comment]
Thanks. I don't seem to have any `.pubxml` files in my project. Perhaps it needs to be created?

**Comment by Toubster** [comment]
Hello Guys,

I am having issues when I try this on a Windows Server 2019.
I also tried it using Windows 10 and a Windows Server 2012R2 and it worked fine.
I tried all the methods i found in this issue and in the issue #7886.
What else can I try to make it work on Windows Server 2019? 

Thanks in advance.

**Comment by ZedTR** [comment]
I'm having this issue after publishing on the client side using IIS6 to host the web application (asp.net core), I tried everything mentioned above and in other threads but no luck, does anyone have any suggestions? please...

**Comment by iSazonov** [comment]
What is SDK version you use?

**Comment by Toubster** [comment]
I use the newest ASP.Net Core SDK 3.1.201.

**Comment by ZedTR** [comment]
@iSazonov the powershell sdk 6.2 its the one that have microsoft.managemnet.infrastructue 1.0.0, if I upgrade to the latest it switches microsoft.management to 2.0.0.
@Toubster I'm using the latest core 3.1.

**Comment by Toubster** [comment]
@ZedTR Are you using the Powershell SDK on a Server 2019? If yes, which version have you Installed from the components on the Server and from the Nuget packets?

**My components versions:**
![grafik](https://user-images.githubusercontent.com/53408989/79075280-2ef59e00-7cf2-11ea-94f0-9b7a449a8bf6.png)


**My Nuget packets:**
![grafik](https://user-images.githubusercontent.com/53408989/79075257-053c7700-7cf2-11ea-8d9d-0a635e62e5e5.png)

**Comment by ZedTR** [comment]
I'm using it on server 2016, I have PowerShell.sdk, Automation and infrastructure downloaded not the diagnostics or the run time installed. not sure about the components on the sever though! can you please tell me how can I get to see what is installed on my server?

**Comment by iSazonov** [comment]
Get-Package -ProviderName NuGet

Please use latest PowerShell SDK version.
I guess you have to add the MMI dll manually into your app publish folder.

**Comment by ZedTR** [comment]
@iSazonov I'm using the latest sdk 7.0.0. I also made sure that I have the management package in the application folder but still no luck. I started to suspect that I might need something called Microsoft Management Framework updated or installed on the server.

**Comment by iSazonov** [comment]
What is targeting in your project?

**Comment by ZedTR** [comment]
.NET core 3.1

**Comment by iSazonov** [comment]
The version has a dll loading tracing. You could use the feature for troubleshooting. I don't know whether this works in IIS.

>  I also made sure that I have the management package in the application folder but still no luck. 

Do you mean that you see mmi dll in your application folder?

Can you share the exception with stack trace?

**Comment by tfriesner** [comment]
I to am running into this issue on Windows Server 2019 with .NET Core 3.1.  Except I am running it from a stand alone service.  Here is the exception I get:

System.IO.FileNotFoundException: Could not load file or assembly  'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral,  PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.  File name: 'Microsoft.Management.Infrastructure, Version=1.0.0.0,  Culture=neutral, PublicKeyToken=31bf3856ad364e35' at  System.Management.Automation.Runspaces.AsyncResult.EndInvoke() at  System.Management.Automation.PowerShell.EndInvoke(IAsyncResult asyncResult) at  System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2  endFunction, Action`1 endAction, Task`1 promise, Boolean  requiresSynchronization) --- End of stack trace from previous location where  exception was thrown
--

I find it interesting that is looking for the 1.0.0.0 version while I have the 2.0.0.0 mmi in my project.  The build will not add the mmi.dll into the application folder and manually adding it will not fix the issue.  I assume its a versioning problem.  How does it determine which version of mmi to load?

As a side note this worked as recently as a week ago. I think the problem started after either a recent Visual Studio 2019 or a powershell SDK (????) update.

**Comment by DivyashriVR** [comment]
I am invoking PoweShell script from .Net Core Console application when I debug the code the PoweShell command is working but when I run the published EXE I am getting below error:
There is exceptionSystem.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.
File name: 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Reflection.Assembly.get_ExportedTypes()
   at System.Management.Automation.Runspaces.PSSnapInHelpers.GetAssemblyTypes(Assembly assembly, String name)
   at System.Management.Automation.Runspaces.PSSnapInHelpers.AnalyzeModuleAssemblyWithReflection(Assembly assembly, String name, PSSnapInInfo psSnapInInfo, PSModuleInfo moduleInfo, String helpFile, Dictionary`2& cmdlets, Dictionary`2& aliases, Dictionary`2& providers)
   at System.Management.Automation.Runspaces.PSSnapInHelpers.AnalyzePSSnapInAssembly(Assembly assembly, String name, PSSnapInInfo psSnapInInfo, PSModuleInfo moduleInfo, Dictionary`2& cmdlets, Dictionary`2& aliases, Dictionary`2& providers, String& helpFile)
   at System.Management.Automation.Runspaces.InitialSessionState.ImportPSSnapIn(PSSnapInInfo psSnapInInfo, PSSnapInException& warning)
   at System.Management.Automation.Runspaces.InitialSessionState.CreateDefault()
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace(PSHost host)
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace()
   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke()
   at EncryptionExe.Program.Main(String[] args).

Is there any solution for this?

**Comment by iSazonov** [comment]
@DivyashriVR Please use Search to find workarounds in the repo. If you use SDK you should "manually" copy the dll to publish directory.

**Comment by DivyashriVR** [comment]
> @DivyashriVR Please use Search to find workarounds in the repo. If you use SDK you should "manually" copy the dll to publish directory.

Hi @iSazonov , I am getting the mentioned error when I run the the trimmed self-contained executable and here I have only 2 files getting generated as in below screenshot:
<img width="528" alt="trimmed self-contained single executable" src="https://user-images.githubusercontent.com/49286033/88261139-58579a00-cce3-11ea-89ee-2c2807ff2fbd.PNG">.

I have created this self-contained executable with the purpose that I don't want any DLL's to be included in it.

So Is there a way to resolve this issue w.r.t self-contained executable.

Thanks

**Comment by iSazonov** [comment]
@DivyashriVR I think you need to put MMI.dll near the exe and look a result.

**Comment by DivyashriVR** [comment]
Hi @iSazonov ,I don't want any dll to be included in my trimmed self-executable.So Is there any way if we can resolve this without adding any dll's to the executable?

**Comment by iSazonov** [comment]
If you put MMI.dll near the exe and this resolve your issue then you can try to modify your csproj so that include the dll in exe. I hope it is possible.

**Comment by OndrejValenta** [comment]
Solution for me taken from #7886 was to run *publish* command manually with **win10-x64**. Visual Studio 2019 in the latest update did not included proper libraries into the publis folder next to the exe file.

```
dotnet publish -o .\publish -r win10-x64
```

**Comment by rodneykahane** [comment]
> 
> 
> > @c01danjo can you explain what you did exactly, if you can still recall?
> 
> I changed the property "RuntimeIdentifier" to "win10-x64" in the publishing configuration. /Properties/{your publishing configuration}.pubxml.

just wanted to chime in for anyone else searching this error, changing the RuntimeIdentifier to from "win-x64" to "win10-64" resolved the **System.IO.FileNotFoundException** error i was receiving.

**Comment by skrubbeltrang** [comment]
I was also able to solve this after a whole day struggling with this issue: Changed the RunTimeIdentifier attribute in pubxml from "win-x64" to "win10-x64".
It was really hard to identify this issue as everything was running smoothly in my application while debugging, and also (mostly when running the published code). Only saw this problem with execution of PowerShell commands in the published code. I used the implicit (default) Runspace in PowerShell and no Exception was thrown before I implemented an explicit Runspace. Then PowerShell automation module generated the error which finally led me to this page:
Could not load file or assembly 'Microsoft.Management.Infrastructure, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified

What an utter waste of a full days work... :-(
